### PR TITLE
Feat: sync application from CR to data store

### LINKED ISSUE #3190: [Feature] Applications synchronization between cli and velaux
**Is your feature request related to a problem? Please describe.**

At the moment, vela cli can manage applications created by velaux, but not the other way around. 

**Describe the solution you'd like**

 Applications created by cli should be visible in velaux.

**Describe alternatives you've considered**

Both cli and velaux should operate through apiserver.  Apiserver pod should bundle with vela-core instead of with velaux. 
**Additional context**

### PR DESCRIPTION
**Pull request by wonderflow**
Signed-off-by: Jianbo Sun <jianbo.sjb@alibaba-inc.com>

### Description of your changes

Fixes #3190

This PR sync Application CR  with velaux datastore

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/apiserver/sync/worker.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		logrus.Fatal(err)
+	}
+
+	f := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dynamicClient, 0, v1.NamespaceAll, nil)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	go startAppSyncing(stopCh, f, ds)
+
+	sigterm := make(chan os.Signal, 1)
```
Comment (barnettZQG) [inline pkg/apiserver/sync/worker.go:54]: There are signal listeners in the main function, so using context is better?

File: pkg/apiserver/sync/crd.go [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+	// This is a double check to make sure the app not be converted and un-deployed
+	sot = CheckSoTFromAppMeta(ctx, ds, targetApp.Name, sot)
+
+	switch sot {
+	case FromUX, FromInner:
+		// we don't sync if the application is not created from CR
+		return nil
+	default:
+	}
+
```
Comment (barnettZQG) [inline pkg/apiserver/sync/crd.go:420]: If the app is already hosted, it cannot be deleted.

File: pkg/utils/strings.go [hunk 3]
```diff
@@ -25,3 +30,33 @@ func StringsContain(items []string, source string) bool {
 	}
 	return false
 }
+
+// CompareSlices will compare two string slice, with the three return values: [both in A and B], [only in A], [only in B]
+func CompareSlices(a []string, b []string) ([]string, []string, []string) {
```
Comment (Somefive) [inline pkg/utils/strings.go:35]: CompareSlices is a bit ambiguous. Maybe ThreeWaySliceCompare or SliceIntersect?

File: pkg/apiserver/sync/crd.go [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+	dsApp, err := ConvertApp2DatastoreApp(ctx, cli, targetApp)
+	if err != nil {
+		log.Logger.Errorf("Convert App to data store err %v", err)
+		return err
+	}
+
+	if err = StoreAppMeta(ctx, dsApp, ds); err != nil {
+		log.Logger.Errorf("Store App Metadata to data store err %v", err)
```
Comment (Somefive) [inline pkg/apiserver/sync/crd.go:474]: What if any one of the StoreXXX fails? Partial information is stored but others are not. Should there be rollback? Or is it idempotent?

File: pkg/apiserver/rest/rest_server.go [hunk 5]
```diff
@@ -117,6 +118,8 @@ func (s *restServer) Run(ctx context.Context) error {
 		leaderelection.RunOrDie(ctx, *l)
 	}()
 
+	go velasync.Start(s.dataStore)
```
Comment (barnettZQG) [inline pkg/apiserver/rest/rest_server.go:121]: Multiple synchronizers work when multiple instances run. so there is a leader election in line 118.

File: pkg/apiserver/sync/crd.go (hunk 2 above)
Comment (wonderflow) [reply pkg/apiserver/sync/crd.go:517]: what do you mean by `hosted`?

Comment (barnettZQG) [reply pkg/apiserver/sync/crd.go:517]: The user manually confirms that VelaUX takes over the application, and only publishing takes place from VelaUX.

Comment (wonderflow) [reply pkg/apiserver/sync/crd.go:517]: Yes, I'm planning to offer an API for that.

File: pkg/velaql/providers/query/collector.go [hunk 6]
```diff
@@ -144,6 +144,9 @@ func (c *AppCollector) FindResourceFromResourceTrackerSpec(app *v1beta1.Applicat
 			}
 			return nil, err
 		}
+		if objRef.Cluster == "" {
+			objRef.Cluster = "local"
```
[the remark this answers is not in the document]
Comment (wonderflow) [reply pkg/velaql/providers/query/collector.go:148]: yes, good suggestion

File: pkg/utils/strings.go (hunk 3 above)
Comment (wonderflow) [reply pkg/utils/strings.go:35]: ThreeWaySliceCompare wins

File: pkg/apiserver/rest/rest_server.go (hunk 5 above)
Comment (wonderflow) [reply pkg/apiserver/rest/rest_server.go:121]: yes, thanks for reminder

File: pkg/apiserver/sync/worker.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/apiserver/clients"
+	"github.com/oam-dev/kubevela/pkg/apiserver/datastore"
+)
+
+// Start prepares watchers and run their controllers, then waits for process termination signals
+func Start(ds datastore.DataStore) {
+
```
[the remark this answers is not in the document]
Comment (wonderflow) [reply pkg/apiserver/sync/worker.go:41]: yes, you're right, I'm going to prevent that

File: pkg/apiserver/sync/worker.go (hunk 1 above)
Comment (wonderflow) [reply pkg/apiserver/sync/worker.go:54]: yes

File: pkg/apiserver/sync/worker.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (wonderflow) [reply pkg/apiserver/sync/worker.go:40]: yes, added some cache

**Review by Somefive — APPROVED** [review]

**Review by barnettZQG — APPROVED** [review]
LGTM

**Pull request merged**